zenml-io/zenml
# Code repository

## Describe changes
This PR adds a code repository abstraction as well as the `GitHub` and `GitLab` implementations for this abstractions.
When a registered code repository is present at the source root when running a pipeline, it will be used to
- track the code repository and the current commit for the pipeline run
- (Optionally) download the source code files inside Docker images instead of including them in the image

### `BaseCodeRepository` vs `LocalRepositoryContext`

The code repository concept has two sides to it:
- The user registers a remote repository (e.g. a GitHub repository) and passes credentials such that ZenML can access the files of this repository. This registers a concrete implementation of the `BaseCodeRepository` interface which can later be used to download files of this repository.
- When running a pipeline (e.g. `python run.py`), ZenML needs to be aware if one of the registered code repositories "is active" for the current source files that the user is running their pipeline with, and some other metadata like the current commit of the repository. This requires a different set of functionalities which are covered by the `LocalRepositoryContext` abstraction.

The connection between the two is implemented in the following way: The `BaseCodeRepository` abstraction has a `get_local_context(path: str) -> Optional[LocalRepositoryContext]` method which subclasses need to implement. In this method, the code repository needs to figure out whether it "exists" at that path and return a local repository if that is the case. This separation allows us to have a shared implementation of a `LocalGitRepositoryContext` which can then be reused by the GitHub/GitLab/BitBucket implementations of the remote functionality.

### How to test this PR
The first step to test this PR in action would be to register a code repository with the following CLI command. Before we can do that, we need to create a GitHub personal access token by following the guide [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token).

Once that is done, run the following command:
```bash
zenml code-repository register <NAME> --type=github --owner=<OWNER> --repository=<GITHUB_REPO_NAME> --token=<GITHUB_PERSONAL_ACCESS_TOKEN>
```
(Example values for `owner` and `repository` when registering the zenml repository: `--owner=zenml-io --repository=zenml`)

If you have your registered repository cloned and now run a pipeline within that repository, ZenML will detect it and store the current commit for you (as long as your repository is not dirty at the point of running the pipeline).

To test the file downloading functionality of the code repository, you need to run with a containerized orchestrator as well as a container registry in your stack. If you now run your pipeline in a clean repository, your code files should not be included in the Docker images but instead be downloaded once the step gets executed. 

If you want to test this with a remote server, ping me for the credentials.

### Other changes:
- The global configuration is not included in Docker images anymore and will instead by set using environment variables.
- The `GitHubActionsOrchestrator` and `GitHubSecretsManager` were removed as they were incompatible with the global configuration change of the previous point.
- The `LocalDockerOrchestrator` error is now much more readable.

## Pre-requisites
Please ensure you have done the following:
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).

## Types of changes

- [ ] Other (add details above)
